File: frontend/js/mathsplit.js
/**
 * task-13: перенос длинных цепочек формул на новые строки (PREVIEW).
 *
 * ЗЕРКАЛО mathsplit.cpp (каноническое правило — оно же описано в
 * mathsplit.h). Фикстуры/тесты: tests/testdata/mathsplit_cases.txt,
 * tests/mathsplit.test.js (node) — результат на них должен совпадать
 * побайтно с C++.
 *
 * window.splitWideMathChains(latex) вызывается ТОЛЬКО для формул,
 * которые рендерятся как display (\[ … \]) — в них перенос «\\»
 * допустим. Для inline формул (в тексте предложения) НЕ вызывать.
 *
 * Правило:
 *   1) Формула содержит \left или \right — не трогаем (fitWideMath).
 *   2) REL = RUN + (пробелы) + отношение (\Leftrightarrow \Rightarrow
 *      \Leftarrow \iff \to \sim \equiv) + (пробелы) + RUN, где RUN —
 *      >= 2 «\,» подряд. Если REL < 2 — не трогаем.
 *   3) Каждое REL -> перенос «\\» + отношение + тонкий пробел «\,».
 * Остальные группы «\,» (без отношения) НЕ трогаем — это пометки
 * операций сайта (| …).
 */
(function () {
    "use strict";

    var REL_RX =
        /(?:\\,){2,}\s*(\\(?:Leftrightarrow|Rightarrow|Leftarrow|iff|to|sim|equiv))\s*(?:\\,){2,}/g;
    var REL_COUNT_RX = new RegExp(REL_RX.source, "g");

    window.splitWideMathChains = function (latex) {
        if (!latex || latex.length === 0) {
            return latex;
        }
        if (latex.indexOf("\\left") !== -1 || latex.indexOf("\\right") !== -1) {
            return latex;
        }
        var rels = (latex.match(REL_COUNT_RX) || []).length;
        if (rels < 2) {
            return latex;
        }
        return latex.replace(REL_RX, function (match, rel) {
            // «\\» (перенос) + отношение в начале строки + тонкий пробел.
            return "\\\\" + rel + "\\, ";
        });
    };
})();

File: mathsplit.cpp
#include "mathsplit.h"

#include <QRegularExpression>

// task-13: реализация канонического правила (см. mathsplit.h).
// Зеркало: frontend/js/mathsplit.js (поведение должно совпадать
// побайтно на фикстурах tests/testdata/mathsplit_cases.txt).

namespace {

// REL: RUN + отношение (группа 1) + RUN.
// Порядок альтернатив важен: длинные команды ПЕРЕД короткими
// (Leftrightarrow перед Leftarrow).
const char *kRelRx =
    R"((?:\\,){2,}\s*(\\(?:Leftrightarrow|Rightarrow|Leftarrow|iff|to|sim|equiv))\s*(?:\\,){2,})";

int countMatches(const QString &s, const QRegularExpression &rx)
{
    int n = 0;
    auto it = rx.globalMatch(s);
    while (it.hasNext()) {
        it.next();
        ++n;
    }
    return n;
}

} // namespace

QString splitWideMathChains(const QString &latex)
{
    if (latex.isEmpty()) {
        return latex;
    }
    // 1) \left/\right — не трогаем (fitWideMath масштабирует).
    if (latex.contains(QLatin1String("\\left"))
        || latex.contains(QLatin1String("\\right"))) {
        return latex;
    }

    static const QRegularExpression relRx{kRelRx};

    // 2) Меньше трёх шагов — не трогаем.
    if (countMatches(latex, relRx) < 2) {
        return latex;
    }

    // 3) REL -> «\\» + отношение + «\, ».
    // ВНИМАНИЕ: в C++ комментарии, кончающиеся обратным слешем, склеиваются
    // со следующей строкой (line-continuation, фаза 2 препроцессора) — такие
    // комментарии здесь ЗАПРЕЩЕНЫ: они глотают следующий код молча.
    QString out;
    int last = 0;
    auto it = relRx.globalMatch(latex);
    while (it.hasNext()) {
        const QRegularExpressionMatch m = it.next();
        out += latex.mid(last, m.capturedStart() - last);
        out += QLatin1String("\\\\");    // перенос строки (два слеша)
        out += m.captured(1);            // отношение в начале строки
        out += QLatin1String("\\, ");    // тонкий пробел
        last = m.capturedEnd();
    }
    out += latex.mid(last);
    return out;
}
